internal/wire/ack_frame.go

```diff
@@ -9,340 +9,166 @@ import (
-	if missingSequenceNumberDeltaLen == 0 {
-		missingSequenceNumberDeltaLen = 1
-	}
+	frame := &AckFrame{}
 
-	largestAcked, err := utils.GetByteOrder(version).ReadUintN(r, largestAckedLen)
+	largestAcked, err := utils.ReadVarInt(r)
 	if err != nil {
 		return nil, err
 	}
 	frame.LargestAcked = protocol.PacketNumber(largestAcked)
-
-	delay, err := utils.GetByteOrder(version).ReadUfloat16(r)
+	delay, err := utils.ReadVarInt(r)
 	if err != nil {
 		return nil, err
 	}
-	frame.DelayTime = time.Duration(delay) * time.Microsecond
-
-	var numAckBlocks uint8
-	if hasMissingRanges {
-		numAckBlocks, err = r.ReadByte()
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	if hasMissingRanges && numAckBlocks == 0 {
-		return nil, ErrInvalidAckRanges
+	frame.DelayTime = time.Duration(delay*1<<ackDelayExponent) * time.Microsecond
+	numBlocks, err := utils.ReadVarInt(r)
+	if err != nil {
+		return nil, err
 	}
 
-	ackBlockLength, err := utils.GetByteOrder(version).ReadUintN(r, missingSequenceNumberDeltaLen)
+	// read the first ACK range
+	ab, err := utils.ReadVarInt(r)
 	if err != nil {
 		return nil, err
 	}
-	if frame.LargestAcked > 0 && ackBlockLength < 1 {
-		return nil, ErrInvalidFirstAckRange
+	ackBlock := protocol.PacketNumber(ab)
+	if ackBlock > frame.LargestAcked {
+		return nil, errors.New("invalid first ACK range")
 	}
+	smallest := frame.LargestAcked - protocol.PacketNumber(ackBlock)
 
-	if ackBlockLength > largestAcked {
-		return nil, ErrInvalidAckRanges
+	// read all the other ACK ranges
+	if numBlocks > 0 {
+		frame.AckRanges = append(frame.AckRanges, AckRange{First: smallest, Last: frame.LargestAcked})
 	}
-
-	if hasMissingRanges {
-		ackRange := AckRange{
-			First: protocol.PacketNumber(largestAcked-ackBlockLength) + 1,
-			Last:  frame.LargestAcked,
+	for i := uint64(0); i < numBlocks; i++ {
+		g, err := utils.ReadVarInt(r)
+		if err != nil {
+			return nil, err
 		}
-		frame.AckRanges = append(frame.AckRanges, ackRange)
-
-		var inLongBlock bool
-		var lastRangeComplete bool
-		for i := uint8(0); i < numAckBlocks; i++ {
-			var gap uint8
-			gap, err = r.ReadByte()
-			if err != nil {
-				return nil, err
-			}
-
-			ackBlockLength, err = utils.GetByteOrder(version).ReadUintN(r, missingSequenceNumberDeltaLen)
-			if err != nil {
-				return nil, err
-			}
-
-			length := protocol.PacketNumber(ackBlockLength)
-
-			if inLongBlock {
-				frame.AckRanges[len(frame.AckRanges)-1].First -= protocol.PacketNumber(gap) + length
-				frame.AckRanges[len(frame.AckRanges)-1].Last -= protocol.PacketNumber(gap)
-			} else {
-				lastRangeComplete = false
-				ackRange := AckRange{
-					Last: frame.AckRanges[len(frame.AckRanges)-1].First - protocol.PacketNumber(gap) - 1,
-				}
-				ackRange.First = ackRange.Last - length + 1
-				frame.AckRanges = append(frame.AckRanges, ackRange)
-			}
-
-			if length > 0 {
-				lastRangeComplete = true
-			}
-
-			inLongBlock = (ackBlockLength == 0)
+		gap := protocol.PacketNumber(g)
```

Comment:
Maybe introduce a utils.ReadVarPacketNumber?

Reply:
Then we'd need the same for stream IDs and byte counts. Not sure if it's worth it.